lib/render/achievement.php

```diff
@@ -1,127 +1,108 @@
+    ?string $context = null,
 ): string {
-    sanitize_outputs(
-        $achName,
-        $consoleName,
-        $gameName,
-        $achPoints
-    );
+    $id = $achievement;
 
-    $smallBadge = '';
-    $displayable = "$achName";
-    if ($achPoints) {
-        $displayable .= " ($achPoints)";
-    }
+    if (is_array($achievement)) {
+        $id = $achievement['AchievementID'] ?? $achievement['ID'];
 
-    if ($inclSmallBadge) {
-        $achNameAttr = attributeEscape($achName);
-        $smallBadgePath = "/Badge/$badgeName" . ".png";
-        $smallBadge = "<img loading='lazy' width='$smallBadgeSize' height='$smallBadgeSize' src='" . media_asset($smallBadgePath) . "' alt='$achNameAttr' title='$achNameAttr' class='$imgclass'>";
+        if ($label !== false) {
+            $title = $achievement['AchievementTitle'] ?? $achievement['Title'];
+            $points = $achievement['Points'] ?? null;
+            $label = $title . ($points ? ' (' . $points . ')' : '');
+        }
+
+        if ($icon !== false) {
+            $badgeName = is_string($icon) ? $icon : $achievement['BadgeName'] ?? null;
+            $icon = media_asset("/Badge/$badgeName.png");
+        }
 
-        if ($smallBadgeOnly) {
-            $displayable = "";
-        } else {
-            $smallBadge .= ' ';
+        if ($achievement['HardcoreMode'] ?? false) {
+            $iconClass = 'goldimage';
         }
+
+        // pre-render tooltip
+        $tooltip = $tooltip !== false ? $achievement : false;
     }
 
-    return "<div class='inline' onmouseover=\"Tip(loadCard('achievement', $achID))\" onmouseout=\"UnTip()\" >" .
-        "<a href='/achievement/$achID'>" .
-        "$smallBadge" .
-        "$displayable" .
-        "</a>" .
-        "</div>";
+    return avatar(
+        resource: 'achievement',
+        id: $id,
+        label: $label !== false && ($label || !$icon) ? $label : null,
+        link: route('achievement.show', $id),
+        tooltip: is_array($tooltip) ? renderAchievementCard($tooltip) : $tooltip,
+        iconUrl: $icon !== false && ($icon || !$label) ? $icon : null,
+        iconSize: $iconSize,
+        iconClass: $iconClass,
+        context: $context,
+    );
 }
 
-function renderAchievementCard(int $achievementId): string
+function renderAchievementCard(int|string|array $achievement, ?string $context = null): string
 {
-    $achData = [];
-    getAchievementMetadata($achievementId, $achData);
+    $id = is_int($achievement) || is_string($achievement) ? (int) $achievement : ($achievement['AchievementID'] ?? $achievement['ID'] ?? null);
+
+    if (empty($id)) {
+        return __('legacy.error.error');
+    }
+
+    $data = [];
+    if (is_array($achievement)) {
+        $data = $achievement;
+    }
+
+    if (empty($data)) {
+        $data = Cache::store('array')->rememberForever('achievement:' . $id . ':card-data', function () use ($id) {
+            $data = [];
+            getAchievementMetadata($id, $data);
```

Comment:
Do we really want to cache the entire metadata object? It includes stuff like `memAddr` that can be rather large, and aren't needed by the tooltip. Since it's just for the duration of the request, it's probably okay for now, but it's something we should revisit later.

Reply:
Thought about that, too, but I don't think this is an issue for now either.

In the future we might want to think about not overfetching data in the first place by using ORM models that only hold the information required to fulfill the task of presenting data in the tooltip. At that point we'd have relations in place, and probably all moved to the blade templating engine.